src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -536,6 +547,70 @@ AppendShardIdToConstraintName(AlterTableCmd *command, uint64 shardId)
 }
 
 
+static bool
+ExtendRowReferencesWalker(Node *node, RowRefWalkerState *state)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, IndexElem))
+	{
+		IndexElem *indexElem = (IndexElem *) node;
+
+		return raw_expression_tree_walker(indexElem->expr, ExtendRowReferencesWalker,
+										  state);
+	}
+	else if (IsA(node, ColumnRef))
+	{
+		ColumnRef *columnRef = (ColumnRef *) node;
+		ListCell *fieldsCell;
+
+		/*
+		 * Append the shardId to any ColumnRef String values that are
+		 * equal to the relationName.  These are actually ROW(relname)
+		 * references.
+		 */
+		foreach(fieldsCell, columnRef->fields)
```

Comment:
> When I tried to use the bare table name like that (in a plain PostgreSQL + ZomboDB install), I couldn't get it to work. Can you give me repro steps to show that it can? I would have just suggested doing that in the first place except I didn't think it worked in that position in the SQL (again, even in a simple non-Citus setup).

If you wanted to see this work in a non-Citus setup, just look at the [tutorial](https://github.com/zombodb/zombodb/blob/master/TUTORIAL.md#creating-an-index).  It shows how this works.

After reading [the Postgres sources you linked above](https://github.com/postgres/postgres/blob/REL9_5_4/src/backend/parser/parse_expr.c#L517-L539), I now understand why it works.  Based on what @anarazel just said, I think we're operating at (at least) a level before that, so we don't necessarily have enough information to fully resolve these `ColumnRefs`.